### PR TITLE
Improve type inference in `DynSystemParam::downcast()` by making the type parameter match the return value.

### PR DESCRIPTION
**Pull request by chescock**
# Objective

Right now, `DynSystemParam::downcast()` always requires the type parameter to be specified with a turbofish.  Make it so that it can be inferred from the use of the return value, like: 

```rust
fn expects_res_a(mut param: DynSystemParam) {
    let res: Res<A> = param.downcast().unwrap();
}
```

## Solution

The reason this doesn't currently work is that the type parameter is a `'static` version of the `SystemParam` so that it can be used with `Any::downcast_mut()`.  Change the method signature so that the type parameter matches the return type, and use `T::Item<'static, 'static>` to get the `'static` version.  That means we wind up returning a `T::Item<'static, 'static>::Item<'w, 's>`, so constrain that to be equal to `T`.  That works with every `SystemParam` implementation, since they have `T::Item == T` up to lifetimes.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@re0312 are you able to review this?

**Review by alice-i-cecile — COMMENTED** [review]
Can we have a standalone test for type inference here? I think it'll make this much clearer if it breaks due to future changes.

File: crates/bevy_ecs/src/system/system_param.rs [hunk 1]
```diff
@@ -1948,7 +1950,10 @@ impl<'w, 's> DynSystemParam<'w, 's> {
 
     /// Returns the inner system param if it is the correct type.
     /// This consumes the dyn param, so the returned param can have its original world and state lifetimes.
-    pub fn downcast<T: SystemParam + 'static>(self) -> Option<T::Item<'w, 's>> {
+    pub fn downcast<T: SystemParam>(self) -> Option<T>
+    where
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:1954]: This where clause could probably use a doc comment: it's *very* complex.

File: crates/bevy_ecs/src/system/builder.rs [hunk 2]
```diff
@@ -545,11 +545,11 @@ mod tests {
...
                 |mut p0: DynSystemParam, mut p1: DynSystemParam, mut p2: DynSystemParam| {
-                    let local = *p0.downcast_mut::<Local<usize>>().unwrap();
-                    let query_count = p1.downcast_mut::<Query<()>>().unwrap().iter().count();
-                    let _entities = p2.downcast_mut::<&Entities>().unwrap();
                     assert!(p0.downcast_mut::<Query<()>>().is_none());
-                    local + query_count
+                    let local: Local<usize> = p0.downcast_mut().unwrap();
+                    let query: Query<()> = p1.downcast_mut().unwrap();
+                    let _entities: &Entities = p2.downcast_mut().unwrap();
+                    *local + query.iter().count()
```
Comment (bushrat011899) [inline crates/bevy_ecs/src/system/builder.rs:552]: I understand that this change you're proposing allows for the variable type to infer the type parameter on `downcast/_mut`, but this change feels rather pointless. In the case of `query_count`, I'd even say it's a regression to move away from turbofish syntax.

File: crates/bevy_ecs/src/system/system_param.rs [hunk 3]
```diff
@@ -1894,7 +1894,9 @@ unsafe impl<T: ?Sized> ReadOnlySystemParam for PhantomData<T> {}
 ///     assert!(param.is::<Res<A>>());
 ///     assert!(!param.is::<Res<B>>());
 ///     assert!(param.downcast_mut::<Res<B>>().is_none());
-///     let foo: Res<A> = param.downcast::<Res<A>>().unwrap();
+///     let res = param.downcast_mut::<Res<A>>().unwrap();
+///     // The type parameter can be left out if it can be determined from use.
+///     let res: Res<A> = param.downcast().unwrap();
```
Comment (bushrat011899) [inline crates/bevy_ecs/src/system/system_param.rs:1899]: Nice documentation, clearly shows the benefit this PR adds.

**Review by bushrat011899 — CHANGES_REQUESTED** [review]
This is a nice ergonomics change for the API, but I agree it needs a comment explaining the `where` clause since it's quite challenging to parse on first glance.

For consistency, the type signature and function body of `is` should also be updated.

```rust
// Before
pub fn is<T: SystemParam + 'static>(&self) -> bool {
    self.state.is::<ParamState<T>>()
}

// After
pub fn is<T: SystemParam>(&self) -> bool
where
    T::Item<'static, 'static>: SystemParam<Item<'w, 's> = T> + 'static,
{
    self.state.is::<ParamState<T::Item<'static, 'static>>>()
}
```

**Review by re0312 — APPROVED** [review]
I might not be very familiar with this part, but it is nice to see improvements of ergonomics.

File: crates/bevy_ecs/src/system/builder.rs (hunk 2 above)
Comment (chescock) [reply crates/bevy_ecs/src/system/builder.rs:552]: Yeah, I had this here to check that inference actually worked, but @alice-i-cecile's suggestion to make a dedicated test for that means I can put this back.

**Comment by chescock** [comment]
> For consistency, the type signature and function body of `is` should also be updated.

Hmm, I had been thinking that it would be simpler to leave it with the simpler type signature, but I think you're right that being consistent is more clear.  I'll change it to match.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**